File: app/src/main/java/pl/orbitemobile/wspolnoty/data/remote/Mapper.java
/*
 * Copyright (c) 2017. All Rights Reserved. Michal Jankowski orbitemobile.pl
 */
package pl.orbitemobile.wspolnoty.data.remote;

import android.support.annotation.NonNull;

import org.jsoup.Connection;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.io.IOException;

import pl.orbitemobile.wspolnoty.data.entities.Article;

public class Mapper {

    @NonNull
    public Article mapArticle(Connection.Response response) throws IOException {
        Article article = new Article();
        article.setTitle(getTitle(response));
        article.setImgUrl(getImgUrl(response));
        article.setDescription(getDescription(response));
        return article;
    }

    @NonNull
    public Article[] mapArticles(Connection.Response response) throws IOException {
        Elements articlesJsoup = getArticles(response);
        Article[] articles = new Article[articlesJsoup.size()];
        for (Element e : articlesJsoup) {
            articles[articlesJsoup.indexOf(e)] = getArticle(e);
        }
        return articles;
    }

    @NonNull
    public Article[] mapNews(Connection.Response response) throws IOException {
        Elements newsJsoup = getNews(response);
        Article[] articles = new Article[newsJsoup.size()];
        for (Element e : newsJsoup) {
            articles[newsJsoup.indexOf(e)] = getSingleNews(e);
        }
        return articles;
    }

    private String getDescription(Connection.Response response) throws IOException {
        return response.parse().getElementsByAttributeValue("class", "entry-content").html()
                .replaceAll("</p>", "\n")
                .replaceAll("<br>", "\n")
                .replaceAll("<[^>]*>", "")
                .trim()
                .replaceAll("\n ", "\n")
                .replaceAll(" \n", "\n")
                .replaceAll("\n\n\n\n\n\n", "\n")
                .replaceAll("\n\n\n\n\n", "\n")
                .replaceAll("\n\n\n\n", "\n")
                .replaceAll("\n\n\n", "\n")
                .replaceAll("&nbsp;", " ")
                .trim();
    }

    private String getImgUrl(Connection.Response response) throws IOException {
        return response.parse().getElementsByAttributeValue("class", "meta-image").get(0).getElementsByTag("img").get(0).absUrl("src");
    }

    private String getTitle(Connection.Response response) throws IOException {
        return response.parse().getElementsByAttributeValue("class", "entry-title").text();
    }

    private Elements getArticles(Connection.Response response) throws IOException {
        return response.parse().getElementsByAttributeValue("class", "vce-featured");
    }

    private Elements getNews(Connection.Response response) throws IOException {
        return response.parse().getElementsByTag("article");
    }

    private Article getArticle(Element element) {
        String title = getTitle(element);
        String imgUrl = getImgUrl(element);
        String articleUrl = getArticleUrl(element);
        return new Article(title, imgUrl, articleUrl);
    }

    private Article getSingleNews(Element element) {
        String title = getNewsTitle(element);
        String imgUrl = getNewsImgUrl(element);
        String articleUrl = getNewsUrl(element);
        Article article = new Article(title, imgUrl, articleUrl);
        article.setDataCreated(getNewsDataCreated(element));
        return article;
    }

    private String getTitle(Element article) {
        return article.getElementsByAttributeValue("class", "vce-featured-title").text();
    }

    private String getNewsTitle(Element article) {
        return article.getElementsByAttributeValue("class", "entry-title").text();
    }

    private String getNewsImgUrl(Element article) {
        return article.getElementsByAttributeValue("class", "meta-image").get(0).getElementsByTag("img").get(0).absUrl("src");
    }

    private String getNewsDataCreated(Element article) {
        return article.getElementsByAttributeValue("class", "updated").text();
    }

    private String getNewsUrl(Element article) {
        Elements elements = article.getElementsByAttributeValue("class", "entry-title").get(0).getElementsByTag("a");
        return elements.get(0).absUrl("href");
    }

    private String getArticleUrl(Element article) {
        Elements elements = article.getElementsByAttributeValue("class", "vce-featured-link-article");
        return elements.get(0).absUrl("href");
    }

    private String getImgUrl(Element article) {
        Elements elements = article.getElementsByTag("img");
        return elements.get(0).absUrl("src");
    }
}

File: app/src/main/java/pl/orbitemobile/wspolnoty/data/remote/Mapper.kt
/*
 * Copyright (c) 2017. All Rights Reserved. Michal Jankowski orbitemobile.pl
 */
package pl.orbitemobile.wspolnoty.data.remote

import android.util.Log
import org.jsoup.Connection
import org.jsoup.nodes.Element
import org.jsoup.select.Elements
import pl.orbitemobile.wspolnoty.data.entities.Article
import pl.orbitemobile.wspolnoty.data.remote.HtmlParser.HtmlParser

open class Mapper {

    val parser = HtmlParser()

    open fun mapArticle(response: Connection.Response, articleUrl: String): Article {
        val title = getTitle(response)
        val imgUrl = getImgUrl(response)
        val description = getDescription(response)
        val article = Article(title, imgUrl, articleUrl)
        article.description = description
        return article
    }

    open fun mapArticles(response: Connection.Response): Array<Article> {
        val articles = getArticles(response)
        Log.d("Mapper", "Article size: " + articles.size)
        return articles.map { getArticle(it) }.toTypedArray()
    }

    open fun mapNews(response: Connection.Response): Array<Article> {
        val news = getNews(response)
        Log.d("Mapper", news.html())
        Log.d("Mapper", "news size: " + news.size)
        return news.map { getSingleNews(it) }.toTypedArray()
    }

    private fun getDescription(response: Connection.Response) =
            parser.parse(response.parse()
                    .getElementsByAttributeValue("class", "entry-content")
                    .html())

    private fun getImgUrl(response: Connection.Response): String {
        return response.parse().getElementsByAttributeValue("class", "meta-image")[0].getElementsByTag("img")[0].absUrl("src")
    }

    private fun getTitle(response: Connection.Response): String {
        return response.parse().getElementsByAttributeValue("class", "entry-title").text()
    }

    private fun getArticles(response: Connection.Response): Elements {
        return response.parse().getElementsByAttributeValue("class", "vce-featured")
    }

    private fun getNews(response: Connection.Response): Elements {
        return response.parse().getElementsByTag("article")
    }

    private fun getArticle(element: Element): Article {
        val title = getTitle(element)
        val imgUrl = getImgUrl(element)
        val articleUrl = getArticleUrl(element)
        return Article(title, imgUrl, articleUrl)
    }

    private fun getSingleNews(element: Element): Article {
        val title = getNewsTitle(element)
        val imgUrl = getNewsImgUrl(element)
        val articleUrl = getNewsUrl(element)
        val article = Article(title, imgUrl, articleUrl)
        article.dataCreated = getNewsDataCreated(element)
        return article
    }

    private fun getTitle(article: Element): String {
        return article.getElementsByAttributeValue("class", "vce-featured-title").text()
    }

    private fun getNewsTitle(article: Element): String {
        return article.getElementsByAttributeValue("class", "entry-title").text()
    }

    private fun getNewsImgUrl(article: Element): String {
        return article.getElementsByAttributeValue("class", "meta-image")[0].getElementsByTag("img")[0].absUrl("src")
    }

    private fun getNewsDataCreated(article: Element): String {
        return article.getElementsByAttributeValue("class", "updated").text()
    }

    private fun getNewsUrl(article: Element): String {
        val elements = article.getElementsByAttributeValue("class", "entry-title")[0].getElementsByTag("a")
        return elements[0].absUrl("href")
    }

    private fun getArticleUrl(article: Element): String {
        val elements = article.getElementsByAttributeValue("class", "vce-featured-link-article")
        return elements[0].absUrl("href")
    }

    private fun getImgUrl(article: Element): String {
        val elements = article.getElementsByTag("img")
        return elements[0].absUrl("src")
    }
}
